Extracting and adding new data to the IDEX L1

bryan-harter opened the pull request:
# Change Summary

## Overview
* Added 30 or so new fields for information that's needed to parse the trigger
* Removed the logging function, all info that was logged is now going into the CDF
* Converting the FPGA time to unix time 

## New Dependencies
None

## New Files
None

## Deleted Files
None

## Updated Files
<!--List each updated file and add bullets to describe the purpose/function of the updates.
This should be high level, but enough detail for the reviewer to understand what purpose(s) the
updated code in the file is serving-->
- imap_processing/idex/idex_packet_parser.py
   - Adding 30 new fields from the header packet into the CDF
   - Removed the logging statement, instead just logging the trigger information
   - Converting the FPGA epoch time to a unix timestamp 
   -  updated the variable name from "data" to "l1_data" in the PacketParser class
- imap_processing/idex/tests/test_decom.py
  - Updated the unit test to reflect the new changes 
- pyproject.toml
  - Ignoring a new ruff rule for functions that are too long.  I'm not sure how to break up the extraction of the 30 new variables further, it seems to me they all belong in the same function.   

## Testing
Updating the unit test to reflect the new changes, no new tests needed though

bryan-harter committed:
Extracting and adding new data to the IDEX L1

bryan-harter commented:
Oh yeah one controversial thing I did was remove the ruff warning for functions that are too long.  I agree functions should be kept short, but extracting like 30 variables from pretty much right next to each other in the header seems like they all belong together.  I could just break it up into smaller functions, but I'm building that big dictionary that I can just loop through later, I kinda feel like a singular function would be better to keep track of the contents of that dictionary better.

bryan-harter commented:
Ehhh you know what, instead of doing all this stuff with bit shifting and masking, I'm going to leave that to space_packet_parser and modify the xml template.  I'll push some changes later today to modify the XML, and hopefully make the python code shorter

tech3371 commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -90,16 +92,36 @@ def __init__(self, packet_file: str):
+        8  # The number of samples in a "block" of low sample data
+    )
+    NUMBER_SAMPLES_PER_HIGH_SAMPLE_BLOCK = (
+        512  # The number of samples in a "block" of high sample data
```
is "block" like 8-bit chunks?

bryan-harter replied:
There are 8 samples in a block, where a sample is just one unit of data collection.  Some sensors in the instrument take samples every 1/4.0625 microseconds, so a block of this low sample data is about 2 microseconds.  The only time that a block of data matters is in a function below, where we calculate how many samples of data are included before the before the dust impact triggered the instrument. 

For whatever reason, instead of the number of samples before triggering, the header has information about the number of blocks before triggering.

greglucas replied:
Maybe add some of that great description as a docstring here?

tech3371 commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -133,35 +161,335 @@ def __init__(self, header_packet):
+        )
+        trigger_notes_dict[
+            "tof_high_trigger_num_min_1_2"
+        ] = """Minimum number of samples between pulse 1 and 2 for TOF High double
```
I see now. `.shortDescription` would have nice in this file. It's very informative.

bryan-harter replied:
Yeah it would certainly shorten the length of this function!

tech3371 commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -133,35 +161,335 @@ def __init__(self, header_packet):
+
+        return trigger_dict, trigger_notes_dict
 
     def _calc_impact_time(self, packet):
```
could I use this function in my PR to convert epoch time to unix time? or can we create a utils folder under `imap_processing` and move this function there since this seems like a common function that we all could benefit from?

Every packet that we get has time in epoch. It's called SHCOARSE or MET.

bryan-harter replied:
Yes that sounds good! I guess it all must be in seconds before 2012?

tech3371 replied:
I will do that in future PR because I don't want extend both PRs. It should be small change.

bryan-harter committed:
Adding some of the values to the XML, rather than shifting bits in the code
(work in progress, need to do a few more variables)

bryan-harter committed:
Finishing up having space packet parser decom

bryan-harter committed:
Getting rid of unecessary bit masksi8kt9gloh

greglucas commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -27,20 +28,20 @@ class PacketParser:
     Attributes
     ----------
-        data (xarray.Dataset): An object containing all of the relevant data in the file
+        l1_data (xarray.Dataset): An object containing all of the relevant L1 data
```
Do you plan on having other data fields around on this object? If not, I liked your generic `data` from before better so it doesn't say `l1_data.l1_data` repeated.

greglucas commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -90,13 +92,27 @@ def __init__(self, packet_file: str):
+    # Bit masks, spelled out for readability
     FOUR_BIT_MASK = 0b1111
     SIX_BIT_MASK = 0b111111
     TEN_BIT_MASK = 0b1111111111
```
Remove all of these now?

Although it looks like you still have one TEN_BIT_MASK left that you might be able to add to the XTCE as well?

greglucas commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -125,6 +142,11 @@ def __init__(self, header_packet):
+        self.trigger_values_dict, self.trigger_notes_dict = self._get_trigger_dicts(
+            header_packet
+        )
+        logging.debug(f"{self.trigger_values_dict}")  # Log values here in case of error
```
```suggestion
        logging.debug(f"trigger_values_dict:\n{self.trigger_values_dict}")  # Log values here in case of error
```
Something to indicate what this is logging otherwise it might be tough to tie the dictionary to the area of interest.

greglucas commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -133,35 +155,318 @@ def __init__(self, header_packet):
     def _calc_impact_time(self, packet):
         """
-        This calculates the number of seconds since Jan 1 2012
+        This calculates the unix timestamp from the FPGA header information
```
Nothing you can do here, but I do find this odd on their part... I would have thought they wanted to keep track of microseconds since MET or some spacecraft pulse. Otherwise, it seems like this value could drift over time since there is no way of synchronizing it with a "true" time on the ground (unless they are doing this every contact, but that still doesn't seem ideal because it could have some significant drift over 3-day periods if you want accurate timings).

greglucas commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -133,35 +155,318 @@ def __init__(self, header_packet):
+        We are given the number of seconds since Jan 1 2012, we need seconds since 1970.
+        Parameters
+        ----------
+            packet: The IDEX FPGA header packet
```
```suggestion
            packet : `space_packet_parser.ParsedPacket`
                The IDEX FPGA header packet
```
or something like that to reference whatever you're passing in as the type.

greglucas commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -133,35 +155,318 @@ def __init__(self, header_packet):
         """
-        # Number of seconds here
-        seconds_since_epoch = packet.data["SHCOARSE"].derived_value
+        # Number of seconds just Jan 1 2012 here
```
```suggestion
        # Number of seconds since Jan 1 2012
```

greglucas commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -133,35 +155,318 @@ def __init__(self, header_packet):
     def _calc_impact_time(self, packet):
         """
-        This calculates the number of seconds since Jan 1 2012
+        This calculates the unix timestamp from the FPGA header information
```
Ahh, I think I see now. You are actually referencing SHCOARSE below, which is the MET time. So that is what I was thinking for synchronizing with pulses on the S/C.

I would update the below to be: `seconds_since_launch` or `mission_elapsed_time` so it is clear. I _think_ that they have arbitrarily chosen Jan 1 2012 as their launch epoch, but that won't actually be the case for us. It is just their testing frame for now.

bryan-harter replied:
Ahh ok, yeah that wasn't clear to me either.  I thought it was weird they chose seconds since 2012.

greglucas commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -133,35 +155,318 @@ def __init__(self, header_packet):
         Calculate the high sample and low sample trigger times from a header packet
+        Parameters
+        ----------
+            packet : The IDEX FPGA header packet info
```
```suggestion
            packet : ...
                The IDEX FPGA header packet info
```

greglucas commented on imap_processing/idex/idex_packet_parser.py:
```diff
@@ -133,35 +155,318 @@ def __init__(self, header_packet):
+            int
+                The actual trigger time for the low sample rate in microseconds
+            int
+                The actual trigger time for the high sample rate in microseconds
```
```suggestion
            (int, int)
                The actual trigger times for the low and high sample rate in microseconds
```

greglucas commented on pyproject.toml:
```diff
@@ -66,4 +67,4 @@ target-version = "py39"
 # Ignore import sorting for now until lines_after_imports is respected
 # by ruff and we can replace isort
-ignore = ["D203", "D212", "PLR0913", "PLR2004", "S101"]
+ignore = ["D203", "D212", "PLR0913", "PLR2004", "S101", "PLR0915"]
```
Do you still need this after your refactor to XTCE?

Perhaps we can add it to `per-file-ignores` for just this one file instead if you do need it?
https://docs.astral.sh/ruff/settings/#per-file-ignores

greglucas reviewed (commented):
Could you add idex to the list of instruments being documented too?

https://github.com/IMAP-Science-Operations-Center/imap_processing/blob/b1a7b65be84de48c1a748155250ad2c4f3642910/docs/source/reference/index.rst?plain=1#L12

bryan-harter committed:
making changes based on comments

bryan-harter committed:
Merge branch 'dev' into idex_l1_cdfs

bryan-harter committed:
Formatting the file after the upstream sync

bryan-harter commented:
Alright I believe I addressed all comments! I'm going to merge just so I can get the CDF review up, but if there were any outstanding issues that I missed I can make tweaks in the next IDEX review
